### PR TITLE
[EC-14]: Server changes for Collection rows in Vault

### PR DESCRIPTION
**Pull request by jlf0dev**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This is the server side changes for bitwarden/clients#3875. These changes include:
- Adding an endpoint to return available organization Collections with associated Groups and Users
    - For admins, this will return all Collections
    - For managers, this will return available Collections
- Adding DeleteMany for Collections

## Code changes

* **\*OrganizationConnectionCommandTests.cs**: All OrganizationConnectionCommandTests were stored under the wrong folder name. Moved to correct folder.
* **CollectionsController.cs:** Update with new endpoints. Do all permission checks for both changes inside controller
* **CollectionRequestModel.cs:** New bulk delete request model, also added to client repo
* **DeleteCollectionCommand.cs:** New DeleteCollectionCommand. Moves the single delete from CollectionService and adds a bulk delete option.
* **Collection_ReadWithGroupsByUserIdOrganizationId.sql:** This stored procedure is a little complex. I wanted to get all the collections available by a user, and filter them by organization. I also then wanted to return (in a separate query) the groups associated with only those collections returned by the user. I ended up with a Temporary Table variable where I can reference the user collections in both queries, but this might not be the best way to handle it. 

I have added tests for the new functionality in the Controller and Command.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        await _eventService.LogCollectionEventAsync(collection, Enums.EventType.Collection_Deleted);
+    }
+    public async Task DeleteManyAsync(IEnumerable<Guid> collectionIds)
+    {
+
+        var ids = collectionIds as Guid[] ?? collectionIds.ToArray();
+        var collectionsToDelete = await _collectionRepository.GetManyByManyIdsAsync(ids);
+        await this.DeleteManyAsync(collectionsToDelete);
+
+    }
```
Comment (coroiu) [inline src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs:31]: nit: you're missing some whitespaces between methods and between constructor and class variables

File: src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        ICollectionRepository collectionRepository,
+        IEventService eventService)
+    {
+        _collectionRepository = collectionRepository;
+        _eventService = eventService;
+    }
+    public async Task DeleteAsync(Collection collection)
+    {
+        await _collectionRepository.DeleteAsync(collection);
+        await _eventService.LogCollectionEventAsync(collection, Enums.EventType.Collection_Deleted);
```
Comment (coroiu) [inline src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs:22]: question: on row 41 you are calling the method with a `deleteDate`, but on this row you are not, did you miss it? Or maybe I'm the one missing something 😄

File: src/Infrastructure.Dapper/Repositories/CollectionRepository.cs [hunk 3]
```diff
@@ -82,6 +95,87 @@ public async Task<ICollection<Collection>> GetManyByOrganizationIdAsync(Guid org
...
+        using (var connection = new SqlConnection(ConnectionString))
+        {
+            var results = await connection.QueryMultipleAsync(
+                $"[{Schema}].[Collection_ReadWithGroupsAndUsersByOrganizationId]",
+                new { OrganizationId = organizationId },
+                commandType: CommandType.StoredProcedure);
+
+            var collections = (await results.ReadAsync<Collection>()).ToList();
+            var groups = (await results.ReadAsync<CollectionGroup>()).ToList();
+            var users = (await results.ReadAsync<CollectionUser>()).ToList();
```
Comment (coroiu) [inline src/Infrastructure.Dapper/Repositories/CollectionRepository.cs:109]: question: Do we really need to convert theses to lists? Shouldn't it be enough if we convert the query results below to lists? I think I saw a discussion on this somewhere but I'm not sure where it ended up.

Comment (coroiu) [reply src/Infrastructure.Dapper/Repositories/CollectionRepository.cs:109]: Comment is valid for rows 147-139 further down in this file

**Review by coroiu — COMMENTED** [review]
Small comments, otherwise looks good to me!

File: src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs (hunk 1 above)
Comment (jlf0dev) [reply src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs:31]: weird, would've thought `dotnet format` catches that...

File: src/Sql/dbo/Stored Procedures/Collection_DeleteByIds.sql [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        [dbo].[Collection]
+    WHERE
+        [Id] in (SELECT [Id] FROM @Ids)
+    GROUP BY
+        [OrganizationId]
+
+    DECLARE @BatchSize INT = 100
+	
+    -- Delete Collection Groups
+    WHILE @BatchSize > 0
```
Comment (shane-melton) [inline src/Sql/dbo/Stored Procedures/Collection_DeleteByIds.sql:22]: Similar to the question in the EF repository, do we need to delete the `CollectionGroups` if they're automatically deleted when the `Collection` is removed?

File: src/Sql/dbo/Stored Procedures/Collection_ReadWithGroupsAndUsersByUserId.sql [hunk 5]
```diff
@@ -0,0 +1,30 @@
+CREATE PROCEDURE [dbo].[Collection_ReadWithGroupsAndUsersByUserId]
+	@UserId UNIQUEIDENTIFIER
+AS
+BEGIN
+    SET NOCOUNT ON
+	
+    DECLARE @TempUserCollections TABLE(Id UNIQUEIDENTIFIER, OrganizationId UNIQUEIDENTIFIER, Name VARCHAR(MAX), CreationDate DATETIME2(7), RevisionDate DATETIME2(7), ExternalId NVARCHAR(300), ReadOnly BIT, HidePasswords BIT)
```
Comment (shane-melton) [inline src/Sql/dbo/Stored Procedures/Collection_ReadWithGroupsAndUsersByUserId.sql:7]: small nit: Could we possibly put the columns on new line breaks, to make this a bit easier to read?

File: test/Api.Test/Controllers/CollectionsControllerTests.cs [hunk 6]
```diff
@@ -85,4 +86,182 @@ await sutProvider.GetDependency<ICollectionService>()
 
         _ = await Assert.ThrowsAsync<NotFoundException>(async () => await sutProvider.Sut.Put(orgId, collectionId, collectionRequest));
     }
+
+    [Theory, BitAutoData]
+    public async Task GetOrganizationCollectionsWithGroups_NoManagerPermissions_ThrowsNotFound(Organization organization, SutProvider<CollectionsController> sutProvider)
```
Comment (shane-melton) [inline test/Api.Test/Controllers/CollectionsControllerTests.cs:91]: Love that the tests are testing each of the user type/permission combinations!

File: test/Core.Test/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommandTests.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            .Returns(new List<Collection> { collection, collection2 });
+
+        // Act
+        await sutProvider.Sut.DeleteManyAsync(collectionIds);
+
+        // Assert
+        await sutProvider.GetDependency<ICollectionRepository>().Received()
+            .DeleteManyAsync(Arg.Is<IEnumerable<Guid>>(ids => ids.SequenceEqual(collectionIds)));
+
+        await sutProvider.GetDependency<IEventService>().Received().LogCollectionEventAsync(collection, EventType.Collection_Deleted, Arg.Any<DateTime>());
```
Comment (shane-melton) [inline test/Core.Test/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommandTests.cs:49]: Here's an example if you update the event service to use bulk collection events:
https://github.com/bitwarden/server/blob/9ea72832db72bc8056ee61516ec324be343091b8/test/Core.Test/OrganizationFeatures/Groups/DeleteGroupCommandTests.cs#L42-L45

File: util/Migrator/DbScripts/2022-10-24_00_CollectionManagement.sql [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        CU.*
+    FROM
+        [dbo].[CollectionUser] CU
+    INNER JOIN
+        [dbo].[OrganizationUser] OU ON OU.[Id] = CU.[OrganizationUserId]
+    WHERE
+        OU.[OrganizationId] = @OrganizationId
+    
+END
+
```
Comment (shane-melton) [inline util/Migrator/DbScripts/2022-10-24_00_CollectionManagement.sql:81]: I believe you'll need a `GO` command here

**Review by shane-melton — CHANGES_REQUESTED** [review]
Looks pretty good to me! Only noticed on a few performance improvements that could be made (that I also discovered during review of my of Groups PR).

I also had a question about your delete repository methods explicitly deleting `CollectionGroup` entities when I would have thought they would have been automatically deleted when the collection is removed. I could be incorrect in that assumption, but if not, you could probably simplify those methods a bit.

File: src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs [hunk 9]
```diff
@@ -238,6 +338,29 @@ public async Task UpdateUsersAsync(Guid id, IEnumerable<CollectionAccessSelectio
         }
     }
 
+    public async Task DeleteManyAsync(IEnumerable<Guid> collectionIds)
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var collectionGroupEntities = await dbContext.CollectionGroups
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs:346]: Nope, the `CollectionGroup` doesn't cascade deletes from `Collection` (although they do from `Group`). Kind of strange.

File: src/Sql/dbo/Stored Procedures/Collection_DeleteByIds.sql (hunk 4 above)
Comment (jlf0dev) [reply src/Sql/dbo/Stored Procedures/Collection_DeleteByIds.sql:22]: Answered above

File: src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public async Task DeleteManyAsync(IEnumerable<Collection> collections)
+    {
+        await _collectionRepository.DeleteManyAsync(collections.Select(c => c.Id));
+
+        var deleteDate = DateTime.UtcNow;
+        foreach (var collection in collections)
+        {
+            await _eventService.LogCollectionEventAsync(collection, Enums.EventType.Collection_Deleted, deleteDate);
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/OrganizationFeatures/OrganizationCollections/DeleteCollectionCommand.cs:41]: Good idea! Fixed

File: test/Api.Test/Controllers/CollectionsControllerTests.cs (hunk 6 above)
Comment (jlf0dev) [reply test/Api.Test/Controllers/CollectionsControllerTests.cs:91]: 😄

File: src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs [hunk 11]
```diff
@@ -180,6 +180,110 @@ where cg.CollectionId.Equals(id)
...
+    public async Task<ICollection<Tuple<Core.Entities.Collection, CollectionAccessDetails>>> GetManyByOrganizationIdWithAccessAsync(Guid organizationId)
+    {
+        var collections = await GetManyByOrganizationIdAsync(organizationId);
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var groups =
+                from cg in dbContext.CollectionGroups
+                where cg.Collection.OrganizationId == organizationId
+                group cg by cg.CollectionId into g
```
Comment (shane-melton) [inline src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs:192]: Did these `group .. by` statement work for you in EF? I was recently trying this myself in the members work and EF core was complaining about it not being translatable to sql. I was forced to do it in after awating: https://github.com/bitwarden/server/blob/d24f0bbfbc1ee1fbc216d1be7231cfbf796bef28/src/Infrastructure.EntityFramework/Repositories/OrganizationUserRepository.cs#L325-L328

**Review by shane-melton — COMMENTED** [review]
I think it's pretty much all there! I'm only surprised to see the working `group .. by` statement in the EF Core Linq query, this works as expected for you?

If so, then we're good to go! (And I need to re-evaluate my EC-15 server PR...)

File: src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs (hunk 11 above)
Comment (jlf0dev) [reply src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs:191]: That's interesting, did you use the  same `group .. by .. into` syntax? I haven't run into any issues with mine, as long as I selected the new `IGrouping` at the end.

Comment (shane-melton) [reply src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs:191]: Hmm, I'm pretty sure I used the `.. into` syntax, but possibly not or I had made some other mistake. But, if this is working as expected for you, it was likely just a problem I had so this PR should be good to go

**Review by shane-melton — APPROVED** [review]
LGTM!

**Pull request merged**